### PR TITLE
Remove `ProxyFactory`

### PR DESCRIPTION
**Pull request by InsertCreativityHere**
This PR removes the `ProxyFactory` class from Java, to further synchronize Java with C++ and C#.
Next up: removing `RequestHandlerFactory`

### CODE REVIEW

File: java/src/Ice/src/main/java/com/zeroc/Ice/ObjectAdapter.java [hunk 1]
```diff
@@ -1339,13 +1337,13 @@ private ObjectPrx newProxy(Identity ident, String facet) {
   private ObjectPrx newDirectProxy(Identity ident, String facet) {
     // Create a reference and return a proxy for this reference.
     var ref = _instance.referenceFactory().create(ident, facet, _reference, _publishedEndpoints);
-    return _instance.proxyFactory().referenceToProxy(ref);
+    return (ref == null) ? null : new com.zeroc.Ice._ObjectPrxI(ref);
```
Comment (pepone) [inline java/src/Ice/src/main/java/com/zeroc/Ice/ObjectAdapter.java:1340]: The null check is not required. The caller already called `checkIdentity`.

File: java/src/Ice/src/main/java/com/zeroc/Ice/ObjectAdapter.java [hunk 2]
```diff
@@ -1339,13 +1337,13 @@ private ObjectPrx newProxy(Identity ident, String facet) {
...
   }
 
   private ObjectPrx newIndirectProxy(Identity ident, String facet, String id) {
     // Create a reference with the adapter id and return a proxy for the reference.
     var ref = _instance.referenceFactory().create(ident, facet, _reference, id);
-    return _instance.proxyFactory().referenceToProxy(ref);
+    return (ref == null) ? null : new com.zeroc.Ice._ObjectPrxI(ref);
```
Comment (pepone) [inline java/src/Ice/src/main/java/com/zeroc/Ice/ObjectAdapter.java:1346]: ditto for null check.

**Review by pepone — APPROVED** [review]

File: java/src/Ice/src/main/java/com/zeroc/Ice/_ObjectPrxI.java [hunk 3]
```diff
@@ -13,8 +13,7 @@
 
 /** Concrete proxy implementation. */
 public class _ObjectPrxI implements ObjectPrx, java.io.Serializable {
-  // TODO: Only public for 'IceInternal.ProxyFactory', should become protected after refactoring.
-  public _ObjectPrxI(Reference ref) {
+  protected _ObjectPrxI(Reference ref) {
```
Comment (bernardnormier) [inline java/src/Ice/src/main/java/com/zeroc/Ice/_ObjectPrxI.java:16]: You should move the protected constructors after the public methods.

**Review by bernardnormier — APPROVED** [review]

File: java/src/Ice/src/main/java/com/zeroc/Ice/_ObjectPrxI.java (hunk 3 above)
Comment (InsertCreativityHere) [reply java/src/Ice/src/main/java/com/zeroc/Ice/_ObjectPrxI.java:16]: Putting constructors before other methods is a _very_ strong Java convention.
Even if those constructors are less visible than the methods that follow.

ie:

1.) Public constructors
2.) Protected constructors
3.) Private constructors

4.) Public methods
5.) Protected methods
6.) Private methods

Comment (InsertCreativityHere) [reply java/src/Ice/src/main/java/com/zeroc/Ice/_ObjectPrxI.java:16]: ... Well, some people will flip these and do methods first, then constructors.
(Although that's less common to see).

But the important part is that constructors should not be mixed in with methods, even if strictly following visibility ordering would say so.

File: java/src/Ice/src/main/java/com/zeroc/Ice/InputStream.java [hunk 4]
```diff
@@ -1538,15 +1538,13 @@ public ObjectPrx readProxy() {
       throw new MarshalException("cannot unmarshal a proxy without a communicator");
     }
 
-    return _instance.proxyFactory().streamToProxy(this);
+    var ident = com.zeroc.Ice.Identity.ice_read(this);
+    var ref = _instance.referenceFactory().create(ident, this);
+    return (ref == null) ? null : new com.zeroc.Ice._ObjectPrxI(ref);
```
[the remark this answers is not in the document]
Comment (InsertCreativityHere) [reply java/src/Ice/src/main/java/com/zeroc/Ice/InputStream.java]: I added the check for consistency. It does represent a minor change in behavior, but no one should be passing empty identities here anyways.

File: java/src/Ice/src/main/java/com/zeroc/Ice/ObjectAdapter.java (hunk 1 above)
Comment (InsertCreativityHere) [reply java/src/Ice/src/main/java/com/zeroc/Ice/ObjectAdapter.java]: Fixed.

File: java/src/Ice/src/main/java/com/zeroc/Ice/ObjectAdapter.java (hunk 2 above)
Comment (InsertCreativityHere) [reply java/src/Ice/src/main/java/com/zeroc/Ice/ObjectAdapter.java]: Ditto for fixed.

**Pull request merged**